File: src/main/java/com/alex/chat/server/ChatServer.java
package com.alex.chat.server;

import com.alex.chat.server.models.User;
import com.alex.chat.server.service.GroupService;
import com.alex.chat.server.service.ReceiverService;
import com.alex.chat.server.service.SenderService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

@Component
@Slf4j
public class ChatServer {
    private final Executor executor = Executors.newCachedThreadPool();

    private final GroupService groupService;

    private final ReceiverService receiverService;

    private final SenderService senderService;

    @Autowired
    public ChatServer(GroupService groupService, ReceiverService receiverService, SenderService senderService) {
        this.groupService = groupService;
        this.receiverService = receiverService;
        this.senderService = senderService;
    }

    /**
     * Запуск сервера
     */
    public void run() throws IOException {
        ServerSocket server = new ServerSocket(5003);
        log.info("Сервер запущен");
        while (!server.isClosed()) {
            Socket socket = server.accept();
            executor.execute(() -> handleConnection(socket));
        }
    }

    /**
     * Обработка подключения пользователя
     *
     * @param socket - сокет, подключаемого клиента
     */
    private void handleConnection(Socket socket) {
        try {
            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            PrintWriter writer = new PrintWriter(socket.getOutputStream());

            User user = authorizeUser(reader);
            executor.execute(receiverService.createReceiver(reader, user));
            executor.execute(senderService.createSender(writer, user));
        } catch (IOException e) {
            log.error("Проблемы с подключением к клиенту");
        }
    }

    /**
     * Получает от клиента имя пользователя и название группы для добавления
     *
     * @param reader входящий поток для чтения данных от клиента
     * @return сущность пользователя
     * @throws IOException если не удалось получить имя пользователя или название группы от клиента
     */
    private User authorizeUser(BufferedReader reader) throws IOException {
        String userName = readUserName(reader);
        String groupName = readGroupName(reader);
        return groupService.addUserToGroup(userName, groupName);
    }

    private String readUserName(BufferedReader reader) {
        try {
            return reader.readLine();
        } catch (IOException e) {
            log.error("Не удалось получить имя пользователя", e);
            throw new RuntimeException(e);
        }
    }

    private String readGroupName(BufferedReader reader) {
        try {
            return reader.readLine();
        } catch (IOException e) {
            log.error("Не удалось получить имя группы", e);
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/com/alex/chat/server/ChatServer.kt
package com.alex.chat.server

import com.alex.chat.server.model.Group
import com.alex.chat.server.model.Message
import com.alex.chat.server.model.User
import com.alex.chat.server.service.GroupService
import com.alex.chat.server.service.UserService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.io.InputStream
import java.io.OutputStream
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException

@Component
class ChatServer(
    private val groupService: GroupService,
    private val userService: UserService,
) {

    suspend fun run() = coroutineScope {
        val server = ServerSocket(5003)
        while (!server.isClosed) {
            val socket = server.accept()
            launch {
                log.info("Accepting new connection")
                handleConnection(socket)
            }
        }
    }

    private suspend fun handleConnection(socket: Socket) = coroutineScope {
        try {
            val inputStream = socket.getInputStream()
            val outputStream = socket.getOutputStream()

            val (username, groupName) = getUserAuthorities(inputStream)
            val group = groupService.findByGroupName(groupName) ?: groupService.create(groupName)
            val user = userService.create(username, group)
            group.addUser(user)
            logJoiningToGroup(user, group)
            launch {
                log.info("Running new message receiver for user ${user.name} of group ${group.name}")
                runReceiver(inputStream, user, group)
            }
            launch {
                log.info("Running new message sender for user ${user.name} of group ${group.name}")
                runSender(outputStream, user)
            }
        } catch (e: SocketException) {
            log.error("Failed to connect to server", e)
        }
    }

    private suspend fun getUserAuthorities(inputStream: InputStream): Authorities {
        val reader = inputStream.bufferedReader()
        val userName = withContext(Dispatchers.IO) {
            reader.readLine()
        }
        val groupName = withContext(Dispatchers.IO) {
            reader.readLine()
        }
        return Authorities(userName, groupName)
    }

    private suspend fun logJoiningToGroup(user: User, group: Group) {
        val text = "${user.name} добавился в группу ${group.name}"
        val message = Message(text)
        sendMessageToGroup(group, message)
    }

    private suspend fun runReceiver(inputStream: InputStream, user: User, group: Group): Unit =
        channelFlow<String> {
            val reader = inputStream.bufferedReader()
            while (true) {
                send(reader.readLine())
            }
        }
            .catch {
                log.error("Error during receiving message", it)
                user.stopMessaging()
            }
            .flowOn(Dispatchers.IO)
            .onCompletion { log.info("Completed receiving messages from user ${user.name} in group ${group.name}") }
            .collect {
                val message = Message(user.name, it)
                sendMessageToGroup(group, message)
            }

    suspend fun sendMessageToGroup(group: Group, message: Message) =
        group.getUsers().asFlow()
            .collect {
                it.sendMessage(message)
            }

    private suspend fun runSender(outputStream: OutputStream, user: User) {
        val writer = outputStream.bufferedWriter()
        user.messages()
            .onCompletion { log.info("Completed sending message to user ${user.name}") }
            .onEach {
                writer.append(it.toString())
                writer.newLine()
                writer.flush()
            }
            .flowOn(Dispatchers.IO)
            .collect {}
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(ChatServer::class.java)
    }

    private data class Authorities(val username: String, val groupName: String)
}
